modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/GlassSystemMenu.java

```diff
@@ -148,7 +160,33 @@ private void insertMenu(final Menu parent, final MenuBase mb, int pos) {
 
         final FilteredList<MenuItemBase> filteredItems = items.filtered(x -> x.isVisible());
 
-        filteredItems.addListener((ListChangeListener.Change<? extends MenuItemBase> change) -> {
+        ListChangeListener menuItemListener = createListener(glassMenu);
+        filteredItems.addListener(menuItemListener);
+        menuListeners.put(glassMenu, menuItemListener);
+        listenerItems.put(menuItemListener, filteredItems);
+
+        for (MenuItemBase item : items) {
+            if (item instanceof MenuBase) {
+                // submenu
+                addMenu(glassMenu, (MenuBase)item);
+            } else {
+                // menu item
+                addMenuItem(glassMenu, item);
+            }
+        }
+        glassMenu.setPixels(getPixels(mb));
+
+        setMenuBindings(glassMenu, mb);
+
+        if (parent != null) {
+            parent.insert(glassMenu, pos);
+        } else {
+            glassSystemMenuBar.insert(glassMenu, pos);
+        }
+    }
+
+    private ListChangeListener<MenuItemBase> createListener(final Menu glassMenu) {
+        return (ListChangeListener.Change<? extends MenuItemBase> change) -> {
```

Comment:
In this listener, when there are items removed that are `Menu`s, we should also call  `clearMenu`:
```
             if (i >= 0 && menuItemList.size() > i) {
                        glassMenu.remove(i);
+                        Object item = menuItemList.get(i);
+                        if (item instanceof Menu menu) {
+                                clearMenu(menu);
+                        }
                 }
```

Otherwise, they won't be removed from the new maps.